File: src/paketBasis.ts
/**
 * Gemeinsame Bausteine aller Pakettexte: Kopfzeilen, Hinweise, Verteilung und
 * der Verzeichnisbaum.
 *
 * Gegenstück in der GUI: `tauri-app/src-tauri/src/paketbasis.rs`. Zahlen stehen
 * ohne Tausendertrenner, damit beide Oberflächen dieselbe Datei erzeugen.
 */
import { GescannteDatei } from './scanner';
import { SchemaMeta } from '@propsa/core';

/** Kopfzeilen, die in jeder Datei des Pakets stehen. */
export function kopf(titel: string, meta: SchemaMeta, zeilen: string[]): string {
  let text = `# PROPSA – ${titel}\n\n`;
  text += `**Projekt:** ${meta.titel}\n`;
  text += `**Erzeugt:** ${meta.zeitstempel}\n`;
  text += `${zeilen.join('\n')}\n`;
  return text;
}

/**
 * Hinweise zur Vollständigkeit.
 *
 * Mit Fail Loud ist ein Scan entweder vollständig oder es gibt kein Paket.
 * Gemeldet werden nur noch Dateien, die das Dateisystem nicht hergab.
 */
export function hinweise(meta: SchemaMeta): string[] {
  const zeilen: string[] = [];
  if (meta.uebersprungen > 0) {
    zeilen.push(
      `- ${meta.uebersprungen} Datei(en) nicht lesbar oder binär und deshalb übersprungen.`
    );
  }
  return zeilen;
}

/** Hinweisblock als eigener Abschnitt (leer, wenn alles vollständig ist). */
export function hinweisblock(meta: SchemaMeta): string {
  const zeilen = hinweise(meta);
  return zeilen.length === 0 ? '' : `\n${zeilen.join('\n')}\n`;
}

/** Zeilen je Sprache, absteigend. */
export function sprachverteilung(dateien: GescannteDatei[]): Array<[string, number]> {
  const proSprache = new Map<string, number>();
  for (const datei of dateien) {
    proSprache.set(datei.sprache, (proSprache.get(datei.sprache) ?? 0) + datei.zeilen);
  }
  return [...proSprache.entries()].sort((a, b) => b[1] - a[1]);
}

/** Anteil als ganze Prozentzahl. */
export function prozent(teil: number, gesamt: number): string {
  const basis = gesamt > 0 ? gesamt : 1;
  return `${Math.round((teil / basis) * 100)} %`;
}

/** Verzeichnisbaum aus den (sortierten) Pfaden. */
export function baum(dateien: GescannteDatei[]): string[] {
  const zeilen: string[] = [];
  let letzteOrdner: string[] = [];

  for (const datei of dateien) {
    const teile = datei.relativerPfad.split('/');
    const ordner = teile.slice(0, -1);

    let gemeinsam = 0;
    while (gemeinsam < ordner.length && letzteOrdner[gemeinsam] === ordner[gemeinsam]) {
      gemeinsam += 1;
    }
    for (let tiefe = gemeinsam; tiefe < ordner.length; tiefe += 1) {
      zeilen.push(`${'  '.repeat(tiefe)}- ${ordner[tiefe]}/`);
    }
    letzteOrdner = ordner;
    zeilen.push(`${'  '.repeat(ordner.length)}- ${teile[teile.length - 1]}`);
  }

  return zeilen;
}

File: tauri-app/src-tauri/src/paketbasis.rs
//! Gemeinsame Bausteine aller Pakettexte: Kopfzeilen, Hinweise, Verteilung und
//! der Verzeichnisbaum.
//!
//! Gegenstück in der CLI: `src/paketBasis.ts`. Zahlen stehen ohne
//! Tausendertrenner, damit beide Oberflächen dieselbe Datei erzeugen.

use crate::scan::{DateiInfo, ScanErgebnis};
use std::collections::HashMap;

/// Kopfzeilen, die in jeder Datei des Pakets stehen.
pub fn kopf(titel: &str, scan: &ScanErgebnis, zeilen: &[String]) -> String {
    let mut text = format!("# PROPSA – {}\n\n", titel);
    text.push_str(&format!("**Projekt:** {}\n", scan.titel));
    text.push_str(&format!("**Erzeugt:** {}\n", scan.zeitstempel));
    text.push_str(&zeilen.join("\n"));
    text.push('\n');
    text
}

/// Hinweise zur Vollständigkeit.
///
/// Mit Fail Loud ist ein Scan entweder vollständig oder es gibt kein Paket.
/// Gemeldet werden nur noch Dateien, die das Dateisystem nicht hergab.
pub fn hinweise(scan: &ScanErgebnis) -> Vec<String> {
    let mut zeilen: Vec<String> = Vec::new();
    if scan.uebersprungen > 0 {
        zeilen.push(format!(
            "- {} Datei(en) nicht lesbar oder binär und deshalb übersprungen.",
            scan.uebersprungen
        ));
    }
    zeilen
}

/// Hinweisblock als eigener Abschnitt (leer, wenn alles vollständig ist).
pub fn hinweisblock(scan: &ScanErgebnis) -> String {
    let zeilen = hinweise(scan);
    if zeilen.is_empty() {
        String::new()
    } else {
        format!("\n{}\n", zeilen.join("\n"))
    }
}

/// Zeilen je Sprache, absteigend.
pub fn sprachverteilung(dateien: &[&DateiInfo]) -> Vec<(String, usize)> {
    let mut pro_sprache: HashMap<&str, usize> = HashMap::new();
    for datei in dateien {
        *pro_sprache.entry(datei.sprache.as_str()).or_insert(0) += datei.zeilen;
    }
    let mut liste: Vec<(String, usize)> = pro_sprache
        .into_iter()
        .map(|(sprache, zeilen)| (sprache.to_string(), zeilen))
        .collect();
    liste.sort_by(|a, b| b.1.cmp(&a.1));
    liste
}

/// Anteil als ganze Prozentzahl.
pub fn prozent(teil: usize, gesamt: usize) -> String {
    let basis = if gesamt > 0 { gesamt } else { 1 };
    let anteil = ((teil as f64 / basis as f64) * 100.0).round() as usize;
    format!("{} %", anteil)
}

/// Verzeichnisbaum aus den (sortierten) Pfaden.
pub fn baum(dateien: &[&DateiInfo]) -> Vec<String> {
    let mut zeilen: Vec<String> = Vec::new();
    let mut letzte_ordner: Vec<&str> = Vec::new();

    for datei in dateien {
        let teile: Vec<&str> = datei.relativer_pfad.split('/').collect();
        let ordner: Vec<&str> = teile[..teile.len() - 1].to_vec();

        let mut gemeinsam = 0;
        while gemeinsam < ordner.len() && letzte_ordner.get(gemeinsam) == Some(&ordner[gemeinsam]) {
            gemeinsam += 1;
        }
        for tiefe in gemeinsam..ordner.len() {
            zeilen.push(format!("{}- {}/", "  ".repeat(tiefe), ordner[tiefe]));
        }
        let name = teile[teile.len() - 1];
        zeilen.push(format!("{}- {}", "  ".repeat(ordner.len()), name));
        letzte_ordner = ordner;
    }

    zeilen
}
